### ISSUE TITLE
[Bug] Vesktop screenshare does not allow to share Pipewire sources (neither applications nor virtual devices)

### ISSUE THREAD

**Issue by mora-unie-youer** [Vencord/venmic#24]
### Discord Account

mora_unie_youer

### Operating System

NixOS 24.11.20240531.57610d2 (Vicuña)

### Linux Only ~ Desktop Environment

Niri (Wayland)

### What happens when the bug or crash occurs?

Audio Sources in Screenshare window do not show anything but Pulseaudio sources.

![image](https://github.com/Vencord/Vesktop/assets/96782282/26b723e3-4ef8-4f35-af6f-3dd11379eae0)

As an example I can provide MPV, but for example osu!lazer and something else is not being shown too.

(MPV is launched using Pipewire Audio Output)
![image](https://github.com/Vencord/Vesktop/assets/96782282/19f5c837-41ff-4871-b80c-a2ffc284b032)

And if I run MPV in Pulseaudio mode, Vesktop shows more audio sources:
![image](https://github.com/Vencord/Vesktop/assets/96782282/d774d867-fa18-4c98-905c-2f4562dbb86e)

(MPV is launched using Pulseaudio Audio Output)
![image](https://github.com/Vencord/Vesktop/assets/96782282/f0b982ee-50ac-40ca-80a7-aa57ff9e5f8a)

### What is the expected behaviour?

Pipewire audio sources must be listed.

### How do you recreate this bug or crash?

1. Start Pipewire application/source/virtual device
2. Try to start screenshare
3. No Pipewire audio source is being shown in menu

### Debug Logs

```
MESA-INTEL: warning: cannot initialize blitter engine
[arRPC > ipc] listening at /run/user/1000/discord-ipc-0
[arRPC > websocket] listening on 6463
[arRPC > process] started
[66179:0623/181146.986992:ERROR:gl_surface_presentation_helper.cc(260)] GetVSyncParametersIfAvailable() failed for 1 times!
[66179:0623/181154.997377:ERROR:gl_surface_presentation_helper.cc(260)] GetVSyncParametersIfAvailable() failed for 2 times!
[66179:0623/181155.001400:ERROR:gl_surface_presentation_helper.cc(260)] GetVSyncParametersIfAvailable() failed for 3 times!
'loop->recurse > 0' failed at ../src/pipewire/thread-loop.c:426 pw_thread_loop_wait()
[2024-06-23 18:11:56.077] [venmic] [info] [patchbay] (handle) found default metadata: 42
[2024-06-23 18:11:56.077] [venmic] [info] [patchbay] (meta_update) speaker name: "alsa_output.pci-0000_30_00.6.analog-stereo"
[2024-06-23 18:11:56.077] [venmic] [info] [patchbay] (get) running venmic 6.1.0
(node:66144) UnhandledPromiseRejectionWarning: TypeError: Video was requested, but no video stream was provided
    at AsyncFunction.<anonymous> (VCDMain:24:2433)
(Use `electron --trace-warnings ...` to show where the warning was created)
(node:66144) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 1)
[66144:0623/181214.226898:ERROR:screencast_portal.cc(367)] Failed to start the screen cast session.
[66144:0623/181214.226926:ERROR:base_capturer_pipewire.cc(81)] ScreenCastPortal failed: 2
Error during screenshare picker Failed to get sources.
(node:66144) UnhandledPromiseRejectionWarning: TypeError: Video was requested, but no video stream was provided
    at AsyncFunction.<anonymous> (VCDMain:24:2180)
(node:66144) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 2)
'loop->recurse > 0' failed at ../src/pipewire/thread-loop.c:426 pw_thread_loop_wait()
(node:66144) UnhandledPromiseRejectionWarning: TypeError: Video was requested, but no video stream was provided
    at AsyncFunction.<anonymous> (VCDMain:24:2433)
(node:66144) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 3)
```

### Request Agreement

- [X] I have searched the existing issues and found no similar issue
- [X] I am using the latest Vesktop and Vencord versions
- [X] This issue occurs on an official release (not just the AUR or Nix packages)

**Comment by Curve** [comment]
Please provide a pw-dump. The pipewire only client is likely missing important node properties

**Comment by mora-unie-youer** [comment]
[pw-dump.log](https://github.com/user-attachments/files/15956252/pw-dump.log)

**Comment by Curve** [comment]
> [pw-dump.log](https://github.com/user-attachments/files/15956252/pw-dump.log)
> 

(haven't looked into the dump yet) But make sure to send the dump when mpv is running in pipewire mode

**Comment by mora-unie-youer** [comment]
Yeah, of course I made dump with MPV in pipewire mode

**Comment by Curve** [comment]
> Yeah, of course I made dump with MPV in pipewire mode

Awesome, will take a look soon!

**Comment by Curve** [comment]
Sorry for the late response.

Granular selection is the culprit in your case, as the native pipewire-node of mpv does not include the PID (which is used for granular selection):

```
      "props": {
        "adapt.follower.spa-node": "",
        "application.icon-name": "mpv",
        "application.id": "mpv",
        "application.name": "mpv",
        "audio.adapt.follower": "",
        "client.id": 208,
        "clock.quantum-limit": 8192,
        "factory.id": 8,
        "factory.mode": "split",
        "library.name": "audioconvert/libspa-audioconvert",
        "media.category": "Playback",
        "media.class": "Stream/Output/Audio",
        "media.name": "[SubsPlease] Spy x Family - Part 2 - 05 [1080p].mkv - mpv",
        "media.role": "Movie",
        "media.type": "Audio",
        "node.always-process": true,
        "node.autoconnect": true,
        "node.description": "mpv",
        "node.name": "mpv",
        "node.rate": "1/44100",
        "node.want-driver": true,
        "object.id": 185,
        "object.register": false,
        "object.serial": 3308,
        "stream.is-live": true
      },
      ```

Disabling granular selection should work in your case, as the node features all the required default props (node name, application name)

**Comment by AizuddinAkmal** [comment]
https://github.com/Vencord/Vesktop/issues/749

I filed a bug earlier in the vesktop repo but found this post in the venmic repo. I am encountering a similar issue with my application (looking-glass) not being populated in the audio sources list. I've tried the workaround by disabling granular selection but the application is still not being populated. I've attached my pw-dump log with audio playing in looking-glass for reference.

[pw-dump.log](https://github.com/user-attachments/files/16192716/pw-dump.log)
